compliance/collection/nodeinventorizer/nodeinventory.go

```diff
@@ -58,23 +58,42 @@ func protoComponentsFromScanComponents(c *nodes.Components) *scannerV1.Component
 	return protoComponents
 }
 
-func convertRHELComponents(rc *database.RHELv2Components) []*scannerV1.RHELComponent {
+func convertAndDedupRHELComponents(rc *database.RHELv2Components) []*scannerV1.RHELComponent {
 	if rc == nil || rc.Packages == nil {
 		log.Warn("No RHEL packages found in scan result")
 		return nil
 	}
-	v1rhelc := make([]*scannerV1.RHELComponent, 0, len(rc.Packages))
-	for _, rhelc := range rc.Packages {
-		v1rhelc = append(v1rhelc, &scannerV1.RHELComponent{
-			// TODO(ROX-13936): Find out if ID field is needed here
+
+	convertedComponents := make(map[string]*scannerV1.RHELComponent, 0)
+	for i, rhelc := range rc.Packages {
+		comp := &scannerV1.RHELComponent{
+			// The loop index is used as ID, as this field only needs to be unique for each NodeInventory result slice
+			Id:          int64(i),
 			Name:        rhelc.Name,
 			Namespace:   rc.Dist,
 			Version:     rhelc.Version,
 			Arch:        rhelc.Arch,
 			Module:      rhelc.Module,
 			Cpes:        rc.CPEs,
 			Executables: rhelc.Executables,
-		})
+		}
+		compKey := makeComponentKey(comp)
+		if compKey != "" {
+			if _, contains := convertedComponents[compKey]; !contains {
+				log.Debugf("Adding component %v to convertedComponents", comp.Name)
+				convertedComponents[compKey] = comp
+			} else {
+				log.Warnf("Detected package collision in Node Inventory scan. Skipping package %s at index %d", compKey, i)
+			}
+		}
+
+	}
+	return maps.Values(convertedComponents)
+}
+
+func makeComponentKey(component *scannerV1.RHELComponent) string {
+	if component == nil {
+		return ""
```

Comment:
I don't get this validation.

I think you are trying to make the code resilient to bad input, which is good. But looking at the caller, we don't anticipate the component to be `nil`. There is no path in the logic that depends on input that might yield a `nil` component either.

I think people reading calls to this function will always think, "Which components will generate empty keys?" to understand "Why are we skipping some components?", and this question is completely irrelevant at the moment.

Even if we could have `nil` components when calling `makeComponentKey()` -- let's say we reuse it in another place where that will be the case -- I would still say it is more readable to leave to the caller the responsibility of checking the condition and dealing with it instead of delegating the check to the key generation function.

Reply:
@Maddosaurus Since this is not a blocker, and everything else LGTM, I decided to approve, and you can either follow up on this here or another time.

Reply:
You're right that it's highly unlikely that we ever receive a `nil` here, as we would most likely have crashed before already.
Also, a very valid point what people reading this function might think. I'll remove the check here to improve readability.

That said, it's an interesting situation, as in an earlier PR review Misha suggested to move `nil` checks further down in the call stack (e.g. let the receiving function handle them).

Reply:
I'll make sure to make my point and opinion clear.

If the function's success depends on factors beyond the programmer's control, it should handle the bad scenarios and signal the caller, usually with status, to let them decide what to do. Otherwise, the function should panic as a sign of a bug in the calling code, a precondition of the function that was not met, as it should never happen in well-written programs. In other words, a bug.

In our case, there is nothing outside our control, and as such, `makeComponentKey()` should simply assume the caller will meet its preconditions and not pass `nil` as input.

If the Component struct would depend on conditions beyond the programmer's control (maybe it comes from some I/O) potentially leading to `nil` during normal program execution, then we should handle the `nil`. I agree that we should handle it down in the call stack, but down to the point where there is context to decide what to do with it. Otherwise, you are masking the `nil` with something else -- in our case, returning an empty key.

Note that sometimes the function you call can deal with `nil`. But most of the time, that should be part of the function's contract, and `nil` should be considered valid input for that function.